pkg/ioutil/ioutil.go

```diff
@@ -205,47 +205,55 @@ func CopyAligned(w *os.File, r io.Reader, alignedBuf []byte) (int64, error) {
 	}
 
 	var written int64
-	var err error
 	for {
-		nr, er := r.Read(alignedBuf)
-		if nr == len(alignedBuf) {
-			// Buffer read is aligned with input buffer, proceed to write.
-			nw, ew := w.Write(alignedBuf)
+		buf := alignedBuf
+		if totalSize != -1 {
+			remaining := totalSize - written
+			if remaining < int64(len(buf)) {
+				buf = buf[:remaining]
+			}
+		}
+		nr, err := io.ReadFull(r, buf)
+		if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+			return written, err
+		}
+		eof := err == io.EOF || err == io.ErrUnexpectedEOF
+		buf = buf[:nr]
```

Comment:
We don't need to reduce the buf pointer. What we can do is use `nr` in this form. 

```diff
diff --git a/pkg/ioutil/ioutil.go b/pkg/ioutil/ioutil.go
index bd4d418d3..46195c070 100644
--- a/pkg/ioutil/ioutil.go
+++ b/pkg/ioutil/ioutil.go
@@ -218,11 +218,10 @@ func CopyAligned(w *os.File, r io.Reader, alignedBuf []byte, totalSize int64) (i
                        return written, err
                }
                eof := err == io.EOF || err == io.ErrUnexpectedEOF
-               buf = buf[:nr]
-               if len(buf)%directioAlignSize == 0 {
+               if nr%directioAlignSize == 0 {
                        // buf is aligned for directio write()
                        var nw int
-                       nw, err = w.Write(buf)
+                       nw, err = w.Write(buf[:nr])
                        if nw > 0 {
                                written += int64(nw)
                        }
```

Reply:
@krishnasrinivas let me know if you want to do this change. Rest is fine and we can take this PR.

Reply:
tried to do this ... to be consistent with the below like `nw, err = writeUnaligned(w, buf[:nr])` but i see that current style is simpler and changed the below line to `nw, err = writeUnaligned(w, buf)`

i.e if `buf = buf[:nr]` is done then `nr` need not be referenced for the rest of the code and `buf` will be self contained

Reply:
The reasoning was since `buf[:nr]` is done in else condition, so you could have stick to that pattern.

Reply:
also refactored a bit to take in poorna's suggestion to simplify if/else statement

Reply:
> The reasoning was since `buf[:nr]` is done in else condition, so you could have stick to that pattern.

yes, i changed this to match the other pattern